src/java.base/share/classes/java/util/stream/GathererOp.java

```diff
@@ -0,0 +1,755 @@
+final class GathererOp<T, A, R> extends ReferencePipeline<T, R> {
+    @SuppressWarnings("unchecked")
+    static <P_IN, P_OUT extends T, T, A, R> Stream<R> of(
+            ReferencePipeline<P_IN, P_OUT> upstream,
+            Gatherer<T, A, R> gatherer) {
+        // When attaching a gather-operation onto another gather-operation,
+        // we can fuse them into one
+        if (upstream.getClass() == GathererOp.class) {
+            return new GathererOp<>(
+                    ((GathererOp<P_IN, Object, P_OUT>) upstream).gatherer.andThen(gatherer),
+                    (GathererOp<?, ?, P_IN>) upstream);
+        } else {
+            return new GathererOp<>(
+                    (ReferencePipeline<?, T>) upstream,
+                    gatherer);
+        }
+    }
+
+    /*
+     * GathererOp.NodeBuilder is a lazy accumulator of elements with O(1)
+     * `append`, and O(8) `join` (concat).
+     *
+     * First `append` inflates a growable Builder, the O(8) for `join` is
+     * because we prefer to delegate to `append` for small concatenations to
+     * avoid excessive indirections (unbalanced Concat-trees) when joining many
+     * NodeBuilders together.
+     */
+    final static class NodeBuilder<X> implements Consumer<X> {
+        private final static int LINEAR_APPEND_MAX = 8; // TODO revisit
+        final static class Builder<X> extends SpinedBuffer<X> implements Node<X> {
+            Builder() {
+            }
+        }
+
+        NodeBuilder() {
+        }
+
+        private Builder<X> rightMost;
+        private Node<X> leftMost;
+
+        private boolean isEmpty() {
+            return rightMost == null && leftMost == null;
+        }
+
+        @Override
+        public void accept(X x) {
+            final var b = rightMost;
+            (b == null ? (rightMost = new NodeBuilder.Builder<>()) : b).accept(x);
+        }
+
+        public NodeBuilder<X> join(NodeBuilder<X> that) {
+            if (isEmpty())
+                return that;
+
+            if (!that.isEmpty()) {
+                final var tb = that.build();
+                if (rightMost != null && tb instanceof NodeBuilder.Builder<X>
+                && tb.count() < LINEAR_APPEND_MAX)
+                    tb.forEach(this); // Avoid conc for small nodes
+                else
+                    leftMost = Nodes.conc(StreamShape.REFERENCE, this.build(), tb);
+            }
+
+            return this;
+        }
+
+        public Node<X> build() {
+            if (isEmpty())
+                return Nodes.emptyNode(StreamShape.REFERENCE);
+
+            final var rm = rightMost;
+
+            if (rm != null) {
+                rightMost = null; // Make sure builder isn't reused
+                final var lm = leftMost;
+                leftMost = (lm == null) ? rm : Nodes.conc(StreamShape.REFERENCE, lm, rm);
+            }
+
+            return leftMost;
+        }
+    }
+
+    final static class GatherSink<T, A, R> implements Sink<T>, Gatherer.Downstream<R> {
```

Comment:
Let's use blessed modifiers order:
```suggestion
    static final class GatherSink<T, A, R> implements Sink<T>, Gatherer.Downstream<R> {
```